### PR TITLE
Dataset namespace resolver feature

### PR DESCRIPTION
**Pull request by pawel-big-lebowski**
### Problem

Oftentimes a dataset's namespace is identified by host used to access it. For example, for Kafka clusters, a first element of bootstrap servers is used to create the namespace. This was the initial approach that deserves to improved and enhanced. 

This PR introduces dataset namespace resolving mechanism, which resolves dataset namespaces based on the resolvers configured. The core mechanism is implemented in `openlineage-java` and can be used within Flink and Spark integrations. 

It contains out-of-the-box resolvers like `hostList`, `pattern` or `patternGroup` type presented below:
```
dataset:
  namespaceResolvers:
      kafka-prod:
        type: hostList
        hosts: ['kafka-prod13.company.com', 'kafka-prod15.company.com']
      cassandra-prod:
        type: pattern
        # 'cassandra-prod7.company.com', 'cassandra-prod8.company.com'
        regex: 'cassandra-prod(\d)+\.company\.com'
      test-pattern:
        type: patternGroup
        # 'cassandra-test-7.company.com', 'cassandra-test-8.company.com', 'kafka-test-7.company.com', 'kafka-test-8.company.com'
        regex: '(?<cluster>[a-zA-Z-]+)-(\d)+\.company\.com:[\d]*'
        matchingGroup: "cluster"
```

and the ability to add custom namespace resolver through the Service Loader. More info will be put in the docs.

Related to: #560

Corresponding docs PR: https://github.com/OpenLineage/docs/pull/329

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2024 contributors to the OpenLineage project

### CODE REVIEW

File: client/java/src/main/java/io/openlineage/client/dataset/hostResolver/DatasetHostResolverBuilder.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/*
+/* Copyright 2018-2024 contributors to the OpenLineage project
+/* SPDX-License-Identifier: Apache-2.0
+*/
+
+package io.openlineage.client.dataset.hostResolver;
+
+public interface DatasetHostResolverBuilder {
```
Comment (d-m-h) [inline client/java/src/main/java/io/openlineage/client/dataset/hostResolver/DatasetHostResolverBuilder.java:8]: Maybe a better name would be "Dataset**Namespace**Resolver" to keep up with the already established vocabulary used within OpenLineage?

**Comment by dolfinus** [comment]
Is it possible to update list of resolvers in already running Spark session?

**Comment by pawel-big-lebowski** [comment]
> Is it possible to update list of resolvers in already running Spark session?

Not now. The list is initiated based on the configuration, that is needed to initiate `OpenLineageContext`

File: client/java/src/main/java/io/openlineage/client/utils/DatasetNamespaceCombinedResolver.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
+/*
+/* Copyright 2018-2024 contributors to the OpenLineage project
+/* SPDX-License-Identifier: Apache-2.0
+*/
+
+package io.openlineage.client.utils;
```
Comment (mobuchowski) [inline client/java/src/main/java/io/openlineage/client/utils/DatasetNamespaceCombinedResolver.java:6]: Why not have this class in the package together with rest of the relevant classes?

File: client/java/src/test/resources/config/datasetNamespaceResolver.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+transport:
+  type: console
+dataset:
+  namespaceResolvers:
```
Comment (mobuchowski) [inline client/java/src/test/resources/config/datasetNamespaceResolver.yaml:4]: Looking at those examples, I wonder if we should not add optional field indicating `schema` of the namespace that the resolver would be operating on. 

For example, I would like to specify that `kafka-prod` resolver would work only on `kafka` namespaces, while `cassandra-prod` would work on `cassandra` namespaces.

I think there exists a risk of mixing those up, especially with too wide regex.

File: client/java/src/main/java/io/openlineage/client/dataset/namespaceResolver/DatasetNamespaceResolver.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+/*
+/* Copyright 2018-2024 contributors to the OpenLineage project
+/* SPDX-License-Identifier: Apache-2.0
+*/
+
+package io.openlineage.client.dataset.namespaceResolver;
```
Comment (mobuchowski) [inline client/java/src/main/java/io/openlineage/client/dataset/namespaceResolver/DatasetNamespaceResolver.java:6]: Everywhere in this PR where we use dot case: maybe `namespace.resolver` instead of `namespaceResolver`?

I don't like mixing up the case styles - let's go with regular Java `dot.separated.case` in packages/directory names and Spark/Flink conf.

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/ArgumentParser.java [hunk 5]
```diff
@@ -197,8 +197,6 @@ private static List<String> getJsonPath(String keyPath) {
   }
 
   private static boolean isArrayType(String value) {
-    return value.startsWith(ARRAY_PREFIX_CHAR)
-        && value.endsWith(ARRAY_SUFFIX_CHAR)
-        && value.contains(DISABLED_FACETS_SEPARATOR);
```
Comment (mobuchowski) [inline integration/spark/app/src/main/java/io/openlineage/spark/agent/ArgumentParser.java:202]: Why was this removed?

File: client/java/src/main/java/io/openlineage/client/dataset/namespaceResolver/HostListNamespaceResolverBuilder.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/*
+/* Copyright 2018-2024 contributors to the OpenLineage project
+/* SPDX-License-Identifier: Apache-2.0
+*/
+
+package io.openlineage.client.dataset.namespaceResolver;
+
+public class HostListNamespaceResolverBuilder implements DatasetNamespaceResolverBuilder {
```
Comment (mobuchowski) [inline client/java/src/main/java/io/openlineage/client/dataset/namespaceResolver/HostListNamespaceResolverBuilder.java:8]: Would be awesome to get proper documentation of those in OpenLineage/docs, but I'm sure you've already thought of that.

**Review by mobuchowski — COMMENTED** [review]
@pawel-big-lebowski astounding work. Would love to hear some clarifications though.

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/ArgumentParser.java (hunk 5 above)
Comment (pawel-big-lebowski) [reply integration/spark/app/src/main/java/io/openlineage/spark/agent/ArgumentParser.java:202]: problem with this are single element arrays which do ton contain `;` separator

**Review by mobuchowski — APPROVED** [review]
